File: Leetcode/221/221.cpp
namespace PrefixSum2D {
    vector<vector<int> > pre;
    int n, m;

    void init(const vector<vector<char> > &A) {

        if (A.empty() || A.front().empty()) return;
        n = (int) A.size();
        m = (int) A.front().size();
        decltype(pre)().swap(pre);
        pre.resize(n, vector<int>(m));

        pre[0][0] = A[0][0] - '0';
        for (int i = 1; i < m; ++i)
            pre[0][i] = pre[0][i - 1] + (A[0][i] - '0');
        for (int i = 1; i < n; ++i)
            pre[i][0] = pre[i - 1][0] + (A[i][0] - '0');
        for (int i = 1; i < n; ++i)
            for (int j = 1; j < m; ++j)
                pre[i][j] = pre[i - 1][j] + pre[i][j - 1] - pre[i - 1][j - 1] + (A[i][j] - '0');
    }

    int rangeSum(int r0, int c0, int r1, int c1) {
        if (r0 == 0 && c0 == 0) return pre[r1][c1];
        else if (r0 == 0 && c0 != 0) return pre[r1][c1] - pre[r1][c0 - 1];
        else if (r0 != 0 && c0 == 0) return pre[r1][c1] - pre[r0 - 1][c1];
        return pre[r1][c1] - pre[r0 - 1][c1] - pre[r1][c0 - 1] + pre[r0 - 1][c0 - 1];
    }
}

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        PrefixSum2D::init(matrix);
        int area = 0;
        int maxSize = max(PrefixSum2D::m, PrefixSum2D::n);
        for (int side = maxSize; side >= 1; --side) {
            for (int i = 0; i + side - 1 < PrefixSum2D::n; ++i) {
                for (int j = 0; j + side - 1 < PrefixSum2D::m; ++j) {
                    if (PrefixSum2D::rangeSum(i, j, i + side - 1, j + side - 1) == side * side) {
                        area = side * side;
                        return area;
                    }
                }
            }
        }
        return area;
    }
};

File: Leetcode/221/221.go
func rangeSum(pre [][]int, r1, c1, r2, c2 int) int {
	if r1 == 0 && c1 == 0 {
		return pre[r2][c2]
	} else if r1 == 0 && c1 != 0 {
		return pre[r2][c2] - pre[r2][c1-1]
	} else if r1 != 0 && c1 == 0 {
		return pre[r2][c2] - pre[r1-1][c2]
	}
	return pre[r2][c2] - pre[r1-1][c2] - pre[r2][c1-1] + pre[r1-1][c1-1]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maximalSquare(matrix [][]byte) int {
	n := len(matrix)
	m := len(matrix[0])
	pre := make([][]int, n)
	for i := 0; i < n; i++ {
		pre[i] = make([]int, m)
	}
	pre[0][0] = int(matrix[0][0] - '0')
	for i := 1; i < n; i++ {
		pre[i][0] = int(matrix[i][0]-'0') + pre[i-1][0]
	}
	for i := 1; i < m; i++ {
		pre[0][i] = int(matrix[0][i]-'0') + pre[0][i-1]
	}
	for i := 1; i < n; i++ {
		for j := 1; j < m; j++ {
			pre[i][j] = pre[i][j-1] + pre[i-1][j] - pre[i-1][j-1] + int(matrix[i][j]-'0')
		}
	}
	maxSide := 1
    hasAns := false
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
            for side := min(n - i, m - j); side >= maxSide; side-- {
				if rangeSum(pre, i, j, i+side-1, j+side-1) == side*side {
                    hasAns = true
					if maxSide < side {
						maxSide = side
					}
					break
				}
			}
		}
	}
    if !hasAns {
        return 0
    }
	return maxSide * maxSide
}